File: 0647-palindromic-substrings/0647-palindromic-substrings.py
class Solution:
    def countPalindromesAroundCenter(self, ss: str, lo: int, hi: int) -> int:
        ans = 0
        
        while lo >= 0 and hi < len(ss):
            if ss[lo] != ss[hi]:
                break
                
            # expand around the center
            lo -= 1
            hi += 1
            
            ans += 1
        return ans
    
    def countSubstrings(self, s: str) -> int:
        ans = 0
        
        for i in range(len(s)):
            # odd-length palindromes, single character center
            ans += self.countPalindromesAroundCenter(s, i, i)
            
            # even-length palindromes, consecutive characters center
            ans += self.countPalindromesAroundCenter(s, i, i + 1)
            
        return ans
    
    
            

File: 0647-palindromic-substrings/0647-palindromic-substrings.cpp
class Solution {
private:
    int countPalindromesAroundCenter(string& ss, int lo, int hi){
        int ans = 0;
        
        while (lo >= 0 && hi < ss.size()){
            // check the first and last characters don't match and expand around the center
            if (ss[lo--] != ss[hi++])
                break;
            
            ++ans;
        }
        
        return ans;
    }
    
public:
    int countSubstrings(string s) {
        int ans = 0;
        
        for (int i = 0; i < s.size(); ++i){
            // odd-length of palindromes, single character center
            ans += countPalindromesAroundCenter(s, i, i);
            
            // even-length of palindromes, consecutive characters center
            ans += countPalindromesAroundCenter(s, i, i + 1);
        }
        
        return ans;
    }
};